kep/205-preemption-toleration/README.md

```diff
@@ -0,0 +1,126 @@
+
+On the other hand, from the cluster administrative view, unconditional non-preempted pods are not desirable because there are system-critical, node-critical pods in the cluster. So, the cluster administrator would like to set the minimum priority value for such a priority class.  For example, assume there are three priority classes: system-critical, high and low.  Then, the cluster administrator would like to introduce `low-non-preempted` priority class that can not be preempted by high but can be preempted by system-critical.
+
+```yaml
+  system-critical: 10000
+             high:  9000
+              low:  8000
+low-non-preempted:  8000 # with minimumPreemptablePriority=10000
+                        # (i.e. high can't preempt this, but system-critical can preempt this)
+```
+
+### Guarantee to running at least N minutes even in lower priority
+
+This is a typical use-case in machine learning job. As described above, job would need check-pointing. However, most machine learning jobs are iteration-based process, called _epochs_. So, if the cluster is high load and preemption always happens before the first epoch is finished,  the machine learning jobs can never make any progress and computational resources consumed by the job came to nothing in this case.
+
+So, cluster administrators would like to exempt pods that do not finish the first epoch from preemption even if the job has lower priority. But, it is hard to monitor when the first epoch is finished outside of the job. Thus, a minimum running time is useful practically.
+
+For example, cluster administrator would introduce several priority classes with the same priority value for introducing several minimum running time guarantee categories like below:
+
+```yaml
+    system-critical: 10000
+                low:  8000  # no minimum running time guarantee
+          low-10min:  8000  # 10 min minimum running time guarantee against 8000 < p < 10000
+          low-30min:  8000  # 30 min minimum running time guarantee against 8000 < p < 10000
+```
+
+## Design Details
+
+### Preemption Toleration API
+
+Preemption toleration policy is defined in this struct type.
+
+```go
+// Please note that this does NOT have ObjectMeta 
+// because this will be embedded in annotation in PriorityClass as unnamed object.
+type PreemptionToleration struct {
+	metav1.TypeMeta `json:",inline"`
+
+  // MinimumPreemptablePriority specifies the minimum priority value that can preempt this priority class.
+  MinimumPreemptablePriority *int32 `json:"minimumPreemptablePriority,omitempty"`
+
+  // TolerationSeconds specified how many seconds this priority class can tolerate preemption by priorities lower than MinimumPreemptablePriority.  Null value specifies forever (default).  Duration means the duration from the pod being scheduled to some node.  This value affects only to scheduled pods (no effect on nominated nodes).
```

Comment:
I hope my understanding is correct:

```suggestion
  // TolerationSeconds specifies how long this priority class can tolerate preemption by priorities lower than MinimumPreemptablePriority.
  // Nil or zero value means it's preemptable immediately (default).
  // Non-nil value means the duration from the pod being scheduled to some node.
  // This value affects scheduled pods only (no effect on nominated pods).
```

Reply:
I intended to the same semantics with [`tolerationSeconds` in Pod](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#toleration-v1-core):

> By default, it is not set, which means tolerate the taint forever (do not evict). Zero and negative values will be treated as 0 (evict immediately) by the system.

Thus, I would like to fix it to be a more clear explanation like bleow:

```suggestion
  // TolerationSeconds specifies how long this priority class can tolerate preemption by priorities lower than MinimumPreemptablePriority.
  // Nil value means forever, which is not preempted forever by the priority value lower than MinimumPreeemptable Priority (default)
  // zero and negative value means immediate, which is preempted immediately (it is probably useless)
  // This value affects scheduled pods only (no effect on nominated pods).
```

Reply:
done in https://github.com/kubernetes-sigs/scheduler-plugins/pull/205/commits/b162e8de84fad9532fa2b2b6866e924c51f5e066